src/leap/mail/adaptors/soledad.py

```diff
@@ -0,0 +1,1047 @@
+            idx = ""
+        return "<%s: %s (%s)>" % (self.__class__.__name__,
+                                  idx, self._doc_id)
+
+
+#
+# Message documents
+#
+
+class FlagsDocWrapper(SoledadDocumentWrapper):
+
+    class model(models.SerializableModel):
+        type_ = "flags"
+        chash = ""
+
+        mbox_uuid = ""
+        seen = False
+        deleted = False
+        recent = False
+        flags = []
+        tags = []
+        size = 0
+        multi = False
+
+        class __meta__(object):
+            index = "mbox"
+
+    def set_mbox_uuid(self, mbox_uuid):
+        # XXX raise error if already created, should use copy instead
+        mbox_uuid = mbox_uuid.replace('-', '_')
+        new_id = constants.FDOCID.format(mbox_uuid=mbox_uuid, chash=self.chash)
+        self._future_doc_id = new_id
+        self.mbox_uuid = mbox_uuid
+
+
+class HeaderDocWrapper(SoledadDocumentWrapper):
+
+    class model(models.SerializableModel):
+        type_ = "head"
+        chash = ""
+
+        date = ""
+        subject = ""
+        headers = {}
+        part_map = {}
+        body = ""  # link to phash of body
+        msgid = ""
+        multi = False
+
+        class __meta__(object):
+            index = "chash"
+
+
+class ContentDocWrapper(SoledadDocumentWrapper):
+
+    class model(models.SerializableModel):
+        type_ = "cnt"
+        phash = ""
+
+        ctype = ""  # XXX index by ctype too?
+        lkf = []  # XXX not implemented yet!
+        raw = ""
+
+        content_disposition = ""
+        content_transfer_encoding = ""
+        content_type = ""
+
+        class __meta__(object):
+            index = "phash"
+
+
+class MetaMsgDocWrapper(SoledadDocumentWrapper):
+
+    class model(models.SerializableModel):
+        type_ = "meta"
+        fdoc = ""
+        hdoc = ""
+        cdocs = []
+
+    def set_mbox_uuid(self, mbox_uuid):
+        # XXX raise error if already created, should use copy instead
+        mbox_uuid = mbox_uuid.replace('-', '_')
+        chash = re.findall(constants.FDOCID_CHASH_RE, self.fdoc)[0]
+        new_id = constants.METAMSGID.format(mbox_uuid=mbox_uuid, chash=chash)
+        new_fdoc_id = constants.FDOCID.format(mbox_uuid=mbox_uuid, chash=chash)
+        self._future_doc_id = new_id
+        self.fdoc = new_fdoc_id
+
+
+class MessageWrapper(object):
+
+    # This could benefit of a DeferredLock to create/update all the
+    # documents at the same time maybe, and defend against concurrent updates?
+
+    implements(IMessageWrapper)
+
+    def __init__(self, mdoc, fdoc, hdoc, cdocs=None):
+        """
+        Need at least a metamsg-document, a flag-document and a header-document
+        to instantiate a MessageWrapper. Content-documents can be retrieved
+        lazily.
+
+        cdocs, if any, should be a dictionary in which the keys are ascending
+        integers, beginning at one, and the values are dictionaries with the
+        content of the content-docs.
```

Comment:
What is the benefit of using a dictionary here instead of a list? Or is just because somewhere else is using this format already and it's easier to keep it like that.